lib/internal/modules/esm/get_format.js

```diff
@@ -112,17 +111,16 @@ function getFileProtocolModuleFormat(url, context, ignoreErrors) {
     if (defaultType === 'commonjs') { // Legacy behavior
       if (packageType === 'none' || packageType === 'commonjs') {
         return 'commonjs';
-      }
-      // If package type is `module`, fall through to the error case below
-    } else { // Else defaultType === 'module'
-      if (underNodeModules(url)) { // Exception for package scopes under `node_modules`
-        return 'commonjs';
-      }
-      if (packageType === 'none' || packageType === 'module') {
-        return getFormatOfExtensionlessFile(url);
-      } // Else packageType === 'commonjs'
+      } // Else packageType === 'module'
+      return getFormatOfExtensionlessFile(url);
+    } // Else defaultType === 'module'
+    if (underNodeModules(url)) { // Exception for package scopes under `node_modules`
       return 'commonjs';
     }
+    if (packageType === 'none' || packageType === 'module') {
+      return getFormatOfExtensionlessFile(url);
+    } // Else packageType === 'commonjs'
+    return 'commonjs';
```

Comment:
This part is wrong. we don't need `underNodeModules` check when `packageType === 'module'`.

Reply:
Can you explain? I think we're checking for under node modules when it's default type module?

Reply:
`underNodeModules` is for backward compatibility for `.js` imports because they are currently loaded as CJS.
We don't need it here because extless imports currently throw.

Reply:
The `underNodeModules` check only occurs for `--default-type=module`. In that mode we still want extensionless files under `node_modules` to be treated as CommonJS because they’re almost certainly part of a dependency that assumes CommonJS (no `type` field). For example, `./node_modules/typescript/bin/tsc` should be treated as a CommonJS file.

Reply:
To put it another way, here’s the algorithm to get the format of an extensionless file:

```
--default-type=commonjs (current Node default):

  "commonjs" package scope ("type": "commonjs", or no "type" field):
    format is "commonjs"

  "module" package scope ("type": "module"):
    format is "module"

--default-type=module:

  "commonjs" package scope ("type": "commonjs", or no "type" field under a node_modules folder):
    format is "commonjs"

  "module" package scope ("type": "module", or no "type" field outside of a node_modules folder):
    format is "module"
```

Reply:
The algorithm is correct but implementation doesn't match.
I'd recommend to create `test/fixtures/es-modules/package-type-module/node_modules/dep-with-package-json-that-has-type-module/noext-esm` and cover it with test that its format is recognized as `module` - with and without flag.

Reply:
How’s this? [`fd3537d` (#49974)](https://github.com/nodejs/node/pull/49974/commits/fd3537d821648831756b593f2fac0bb0ccc9e014)

Reply:
Better but still not correct.
We need to use `getFormatOfExtensionlessFile` so wasm is also supported.